components/calendar/src/week_of.rs

```diff
@@ -52,20 +52,20 @@ impl WeekCalculator {
     ///
     /// [📚 Help choosing a constructor](icu_provider::constructors)
     #[cfg(feature = "compiled_data")]
-    pub fn try_new(locale: &DataLocale) -> Result<Self, CalendarError> {
+    pub fn try_new(locale: &Locale) -> Result<Self, CalendarError> {
```

Comment:
Issue: This is a regressive change.

1. We have evidence that `Locale` is bigger and slower than `DataLocale`
2. We are migrating to `Preferences`, and the `.into()` will make that change more likely to "just work" when it is ready to land.

(I thought we were aligned on this. I'm surprised to see this PR in my inbox.)

Reply:
This is a step in this direction. The next step will be to replace `Locale` by `Preferences`, but at some point we have to decouple the API type from the provider type.

Reply:
> 1. We have evidence that Locale is bigger and slower than DataLocale

This only uses `Locale` on the API boundary, I then directly borrow the `LanguageIdentifier` and `DataKeyAttributes`, which should be both smaller and faster than the current model.

Reply:
I changed to title to emphasize the important change, getting `DataRequest` to the design in https://github.com/unicode-org/icu4x/issues/3632

Reply:
So it sounds like you want to keep `DataLocale` on the API so that we don't remove the `into()`s now? I can do that

Reply:
I think my preferred transitional state would be keeping DataLocale and all public call sites the same, adding a function to DataLocale that returns &LanguageIdentifier, and using it when constructing the DataRequest

Reply:
This is basically what I've done, but I've made DataLocale an open struct and deleted all of its logic.